rfcs/2023-12-19-ADMIN_AUTH.md

```diff
@@ -0,0 +1,302 @@
+Dome.fm instance is sending `less than or equal to 5 emails per minute` (more below).
+
+3. Before triggering the email, the Mothership API will perform an authenticity handshake (more below).
+
+4. Before triggering the email, the Mothership API will test that the recipient email is a known
+owner or administrator of the Dome.fm instance.
+
+5. If all security checks pass, the email will be sent, with a body like:
+
+```
+<a href="https://developers.dome.fm/admin/authenticate?token=[token]&instance=[X-DOME-HOSTNAME]">Login to Dome.fm</a>
+```
+
+**Note:** Wait, why does this email body return to `developers.dome.fm` and not `X-DOME-HOSTNAME`?
+
+Well, because we need to know about the existence of our decentralized admin users in the Mothership API
+as a security precaution, we also need to know that they didn't enter their email addresses incorrectly.
+
+So, we "proxy" the magic link completion step through the Mothership API (which simply "confirms" the email
+address in the Mothership database) and then redirects to `X-DOME-HOSTNAME` for the Instance to actually
+issue a valid session.
+
+Ie, this allows a single magic link to be "double confirmable" in both the Mothership AND the Instance,
+and in the future, allows us to issue a `*.dome.fm` cookie for a universal web player and unified login
+experience!
+
+## How security will work
+
+Technically, anyone on the internet can register for a Mothership API key (and thus send emails). So,
+here's a few security approaches we'll take to ensure malicious users can't gain access to an instance
+admin panel.
+
+### Handshaking
+
+Every request to the Mothership API will include a handshake:
+
+1. `artistname.vercel.app` will make some request to `developers.dome.fm/instance-api/...`
+2. That request must include `X-DOME-HOSTNAME: "artistname.vercel.app"` and `X-DOME-REQUEST-ID: "some-uuid-generated-by-instance"` (which will be stored in PostgresQL for ~60 minutes)
+3. `developers.dome.fm` will return a request to `artistname.vercel.app/api/authenticity-handshake?requestId=some-uuid-generated-by-instance`
+4. If the requestId is found in the Instance PostgresQL database, it will return a bodyless `204` response
+5. Upon receiving that `204`, the request will be allowed to proceed further into the stack. (If not the
+request stack will be halted)
+
+### API Keys
+
+1. The request will include headers `X-DOME-INSTANCE-ID` and `X-DOME-MOTHERSHIP-API-KEY`,
+which will be checked for a match in the Mothership PostgresQL instance.
+
+### Rate Limiting
+
+1. The request will be rate limited against the referrer IP address.
+
+### Testing Recipients
+
+On registration, we'll record the instance owner email address in the Mothership API. (This
+is necessary for security alerts in the future).
+
+In order to ensure we only send magic link emails to the appropriate instance admins (and not
+a malicious user), we'll check that the `email` field given is actually a known admin by the
+Mothership API.
```

Comment:
> we'll check that the `email` field given is actually a known admin by the
Mothership API.

Could you add a bit more detail on how we'd perform this check? I assume that the Mothership would store the `owner` included with requests to the `/instances` endpoint. In order to support ownership changes / multiple admins on a single instance, we'll also need to notify the Mothership when there are changes to the instance "team".

Reply:
> Could you add a bit more detail on how we'd perform this check? I assume that the Mothership would store the owner included with requests to the /instances endpoint.

Correct. The `Instance - has_many -> Administrator (role: owner)`

> In order to support ownership changes / multiple admins on a single instance, we'll also need to notify the Mothership when there are changes to the instance "team"

Ya, exactly. Annoying I know but ultimately good for the ecosystem's security.

The instances will need to sync their state against the mothership API periodically. It could be done on a per-request basis, but honestly I'm kinda thinking a better way to do that would be to periodically `pg_dump` the entire instance DB and send it to the mothership. 

Why? That's kinda heavy handed...

Well because if we want to support a player app & centralized discovery (ie, `dome.fm/search-artists`) we'd need to essentially copy the entire DB of the instance anyhow. & If a fan logs into a player app, we'd need to know everything they've bought from various Dome instances... which requires us to aggregate everything into one big collection to power that experience.

(Of course, this periodic sync could be "turned off", ie, you could "undock" your instance completely, but you'd be excluded from centralized discovery)

Reply:
Cool. That makes sense to me. Certainly a reasonable place to start anyway, we could make it more surgical down the line if necessary.